Don't set locale used by C++ stdlib to the value from `LC_ALL` env variable even if it's valid

wechman opened the pull request:
fixes #12350

Commit:
Use "C" instead user environment locale in solc

cameel committed:
A script and CI job for testing the compiler with different locales

cameel commented:
From the description in #674 it looks to me like the reporter had the locale variables in his system set to a locale that was not enabled (i.e. not present or commented-out in `/etc/locale.gen`). The thing I'm not 100% sure about is why it affected solc at all. `setlocale()` docs say that the stdlib does `setlocale(LC_ALL, "C")` at startup, which means that it should completely ignore the locale env vars - so why would invalid locale name in these vars do anything?

The only way I can explain this is that something else down the line tries to use these env variables and fails. Could be Boost or maybe we're spawning a Bash process at some point to run a shell command? If that's the case, it would explain why `setenv()` added by #975 helped - this way we reset the locale variable inherited by the solc process to "C" if its value is invalid. This means that whatever tries to use that variable, now never has to deal with an invalid value.

I hate to be guessing here and given that this locale problem was lurking for several years, I think we should verify that the solution is correct. We need a test that would detect **both** #12350 and #674. This might not be the easiest task if you're not familiar with our CI and scripts so I just tried to create one myself. Here's my [`locale-test`](https://github.com/ethereum/solidity/tree/locale-test) branch. It adds a test script and a CI job that runs it. I'm not submitting it as a PR because it of course fails on `develop` - it detects the problem with the Turkish locale. Please cherry-pick commits from it into your branch and then:
1) Make sure that after your fix it passes.
2) Try removing the `setenv()` line to verify that the test is effective against #674 - if it does not fail after that, the test needs to be improved. If you can't make it fail, it might indicate that we don't really need the `setenv()`.

cameel commented on solc/main.cpp:
```diff
@@ -36,31 +36,21 @@ using namespace solidity;
-	}
+	setenv("LC_ALL", "C", 1);
 #endif
 }
```
The name and docs should be adjusted to more precisely describe what is happening here and why. I think that this function is no longer named correctly. It does not really set the locale used by the std library any more. It just adjusts the locale set in the env variable if it's invalid. This variable is supposed to be ignored by stdlib though.

cameel commented:
Just for reference, here's [a failing run of my new `t_ubu_locale` job](https://app.circleci.com/pipelines/github/ethereum/solidity/22448/workflows/709f21c2-45cd-4c27-bca6-daab095e62d9/jobs/985548), which shows that it at least detects the problem from #12350 on `develop.`

cameel commented:
Oh, one more thing. This fixes a user-visible bug so you should add a new entry in the [changelog](https://github.com/ethereum/solidity/blob/develop/Changelog.md) under "Bugfixes".

wechman commented:
@cameel  Thank you for your tests! They cover all scenarios I carried out in my local environment. Unfortunately, I was not able to reproduce #674 and neither your tests are. "setenv" removal does not change anything in this matter (seems we can remove it).

Regardless of the above, your guessing sounds reasonable to me. I think, the problem might be related to the dependency we no longer use (e.g. particular version of boost or other library). This could explain why we cannot reproduce it.

chriseth commented:
I would actually be happier if we could find a way to somehow set an "illegal" locale so that we get an error whenever we perform a locale-dependent operation.

cameel commented on Changelog.md:
```diff
@@ -7,7 +7,7 @@ Compiler Features:
 
 Bugfixes:
-
+ * The compiler uses "C" instead of environment locale.
```
This describes how we fixed it but I think that in the changelog we should focus on the effects of the bug visible to the users (i.e. it caused an ICE under certain conditions):
```suggestion
 * Fix internal error for locales with unusual capitalization rules. Locale set in the environment is now completely ignored.
```

cameel reviewed (commented):
I think it's good now. I'd just slightly adjust the changelog entry.

cameel commented:
> I would actually be happier if we could find a way to somehow set an "illegal" locale so that we get an error whenever we perform a locale-dependent operation (and of course remove all locale-dependent operations).

Yeah. would be nice if we could detect locale-dependent code. I think that there might be a lot of such code though (pretty much everything related to capitalization for example) so I'd treat it as a separate issue.

chriseth commented:
Also, does it really fix the issue? It only fixes it in one binary...
I think we really need to find all those locale-dependent functions. I don't think there are too many, by the way.

cameel commented:
Ah, I was thinking about `std::locale` but I see that's independent of `std::setlocale()` (and not global).

By the way, the docs for `setlocale()` have a nice list of functions affected by the global state it manages: https://github.com/ethereum/solidity/issues/12720#issuecomment-1054485424. That's over 40 different functions. ~20 if you don't count the ones meant for wide chars.

cameel reviewed (dismissed):
> I would prefer both be implemented :)

If possible, sure. But for now it's not even certain that it's possible to set an invalid locale (https://github.com/ethereum/solidity/issues/12720#issuecomment-1054570799). The part that is done here is self-contained and we want it even if we won't find a way to set an invalid locale so I see no reason not to merge it already.

cameel commented:
@wechman Can you switch it to non-draft and squash the fixup commit?

chriseth commented on test/localeTest.sh:
```diff
@@ -0,0 +1,79 @@
+EOF
+)
+
+# Whatever locale is set by default.
```
Does this test actually fail before the fix?
Can you make it generate some code and check that it is identical for all locales?

cameel replied:
> Does this test actually fail before the fix?

Yes, see https://github.com/ethereum/solidity/pull/12716#issuecomment-1051261952 and https://github.com/ethereum/solidity/pull/12716#issuecomment-1054096720 above.

> Can you make it generate some code and check that it is identical for all locales?

I assumed that with bad locale it would not compile so this would be enough, but true, we cold write output from all `solc` calls to files (instead of sending it to `/dev/null`) and `diff` it at the end.

cameel commented on test/localeTest.sh:
```diff
@@ -0,0 +1,83 @@
+printTask "Testing the C locale..."
+export LC_ALL=C
+[[ ${i^^} == "I" ]] || assertFail
+echo "$test_code" | "$solc_binary" - --bin > c_locale_result.out
+diff_values "$(<default_locale_result.out)" "$(<c_locale_result.out)"
```
Please avoid creating temp files in the working dir. These scripts should be usable also outside of CI and it's not a good idea to assume that the work dir does not have files with these names or even that you can write to it. You can create a dir when you can write safely with:
```bash
tmp_dir=$(mktemp -d -t "locale-test-XXXXXX")
```
And remember to delete it at the end of the script.

Alternatively, you could capture the output to a variable - I think this is usually preferable:
```suggestion
c_locale_output=$("$test_code" | "$solc_binary" - --bin)
diff_values "$(echo "$default_locale_output")" "$(echo "$c_locale_output")"
```

cameel reviewed (commented):
The script should be adjusted not to create temp files in the working dir. Other than that the PR is fine.

wechman commented:
> The script should be adjusted not to create temp files in the working dir. Other than that the PR is fine.

Ok, script updated. I already squashed all commits and rebased on the newest develop.

cameel commented on Changelog.md:
```diff
@@ -11,7 +11,7 @@ Compiler Features:
 Bugfixes:
  * Yul IR Code Generation: Optimize embedded creation code with correct settings. This fixes potential mismatches between the constructor code of a contract compiled in isolation and the bytecode in ``type(C).creationCode``, resp. the bytecode used for ``new C(...)``.
-
+ * Fix internal error for locales with unusual capitalization rules. Locale set in the environment is now completely ignored.
```
Just a minor nitpick, but I'd keep the extra line here to keep spacing consistent.

wechman replied:
Problem fixed.
